src/test/java/com/amazonaws/encryptionsdk/model/KeyBlobTest.java

```diff
@@ -170,4 +171,52 @@ public void checkKeyLen() {
 
         assertEquals(mockDataKey_.getEncryptedDataKey().length, reconstructedKeyBlob.getEncryptedDataKeyLen());
     }
+
+    private KeyBlob generateRandomKeyBlob(int idLen, int infoLen, int keyLen) {
+        final byte[] idBytes = RandomBytesGenerator.generate(idLen);
+        // negative bytes translate into U+FFFD, so no thanks
+        for (int i = 0; i < idBytes.length; i++) {
+            if (idBytes[i] < 0) {
+                idBytes[i] = (byte) (idBytes[i] - Byte.MIN_VALUE);
+            }
+        }
+        final byte[] infoBytes = RandomBytesGenerator.generate(infoLen);
+        final byte[] keyBytes = RandomBytesGenerator.generate(keyLen);
+
+        return new KeyBlob(new String(idBytes, StandardCharsets.UTF_8), infoBytes, keyBytes);
+    }
+
+    private void assertKeyBlobsEqual(KeyBlob b1, KeyBlob b2) {
+        assertArrayEquals(b1.getProviderId().getBytes(StandardCharsets.UTF_8),
+                          b2.getProviderId().getBytes(StandardCharsets.UTF_8));
+        assertArrayEquals(b1.getProviderInformation(), b2.getProviderInformation());
+        assertArrayEquals(b1.getEncryptedDataKey(), b2.getEncryptedDataKey());
+    }
+    
+    @Test
+    public void checkKeyProviderIdLenUnsigned() {
+        // provider id length is too large for a signed short but fits in unsigned
+        final KeyBlob blob = generateRandomKeyBlob(Short.MAX_VALUE + 1, Short.MAX_VALUE, Short.MAX_VALUE);
+        final byte[] arr = blob.toByteArray();
+
+        assertKeyBlobsEqual(deserialize(arr), blob);
+    }
+
+    @Test
+    public void checkKeyProviderInfoLenUnsigned() {
+        // provider info length is too large for a signed short but fits in unsigned
+        final KeyBlob blob = generateRandomKeyBlob(Short.MAX_VALUE, Short.MAX_VALUE + 2, Short.MAX_VALUE);
+        final byte[] arr = blob.toByteArray();
+
+        assertKeyBlobsEqual(deserialize(arr), blob);
+    }
+
+    @Test
+    public void checkKeyLenUnsigned() {
```

Comment:
Can you add a test for when it is too long to fit even in an unsigned short?

Reply:
Good idea; I will

Reply:
It occurs to me now that such tests already exist, namely `overlyLargeKeyProviderIdLen()`, `overlyLargeKeyProviderInfoLen()`, and `overlyLargeKey()`. They check that instantiating such a KeyBlob results in an exception. (I had changed them earlier to make sure they were treating `Constants.UNSIGNED_SHORT_MAX_VAL` as the max length.)